### PR TITLE
Closes #573: add tests for quake.concat.

### LINKED ISSUE #573: ConcatOp Canonicalization
I would like to demonstrate lowering 
```cpp

__qpu__ void cnotKernel(cudaq::qubit &q, cudaq::qubit &r) {
  x<cudaq::ctrl>(q, r);
}

__qpu__ void multiCtrlTest() {
  cudaq::qreg q(4);
  cudaq::control(cnotKernel, {q[0], q[1]}, q[2], q[3]);
}

```

If I run 
```bash 
bin/cudaq-quake demo/multicontrol_decomp.cpp | \
  bin/cudaq-opt --canonicalize --apply-op-specialization  \
      --inline --quake-multicontrol-decomposition \
      --quantinuum-gate-set-mapping --canonicalize
```
you get 
```mlir
func.func @__nvqpp__mlirgen__function_multiCtrlTest._Z13multiCtrlTestv() attributes {"cudaq-entrypoint", "cudaq-kernel"} {
    %0 = quake.alloca !quake.ref
    %1 = quake.alloca !quake.veq<4>
    %2 = quake.extract_ref %1[0] : (!quake.veq<4>) -> !quake.ref
    %3 = quake.extract_ref %1[1] : (!quake.veq<4>) -> !quake.ref
    %4 = quake.concat %2, %3 : (!quake.ref, !quake.ref) -> !quake.veq<2>
    %5 = quake.extract_ref %1[2] : (!quake.veq<4>) -> !quake.ref
    %6 = quake.extract_ref %1[3] : (!quake.veq<4>) -> !quake.ref
    %7 = quake.concat %4 : (!quake.veq<2>) -> !quake.veq<2>
    %8 = quake.extract_ref %7[0] : (!quake.veq<2>) -> !quake.ref
    %9 = quake.extract_ref %7[1] : (!quake.veq<2>) -> !quake.ref
    quake.h %0 : (!quake.ref) -> ()
    quake.x [%9] %0 : (!quake.ref, !quake.ref) -> ()
    quake.t<adj> %0 : (!quake.ref) -> ()
    quake.x [%8] %0 : (!quake.ref, !quake.ref) -> ()
   ... more ...
```

Notice the `quake.concat` operations that remain. They are superfluous and could probably be optimized away. Is there any way to do that? Here we extract qubits 0 and 1, and concat them to a `veq<2>`, then we run `concat` again (looks like a no-op), and then just extract those qubits back out. Ideally, we could have just extracted from the original `veq` and been done.

### PR DESCRIPTION
**Pull request by schweitzpgi**
Adds tests for the canonicalization of quake.concat.

### CODE REVIEW

File: test/Quake/cse.qke [hunk 1]
```diff
@@ -8,39 +8,48 @@
...
+  quake.x [%4] %1   : (!quake.ref, !quake.ref) -> ()
+  return
+}
+
 // CHECK-LABEL: func.func @__nvqpp__mlirgen__deuteron_n3_ansatz2
 // CHECK: cc.stdvec_data
 // CHECK-NOT: cc.stdvec_data
 // CHECK: return
 
-module {
```
Comment (khalatepradnya) [inline test/Quake/cse.qke:16]: Clarifying Q: This was moved but, `module` was removed... Was 'module' redundant to begin with or a new requirement?

File: test/Quake/cse.qke [hunk 2]
```diff
@@ -49,14 +58,84 @@ module {
...
+  %q11 = quake.concat %q10 : (!quake.veq<?>) -> !quake.veq<?>
+  call @device_kernel(%q11) : (!quake.veq<?>) -> ()
+  return
 }
 
+// CHECK-LABEL:   func.func @canonicalize_concat() {
+// CHECK:           %[[VAL_0:.*]] = quake.alloca !quake.ref
+// CHECK:           %[[VAL_1:.*]] = quake.concat %[[VAL_0]] : (!quake.ref) -> !quake.veq<1>
+// CHECK:           %[[VAL_2:.*]] = quake.mz %[[VAL_1]] : (!quake.veq<1>) -> !cc.stdvec<!quake.measure>
+// CHECK:           %[[VAL_3:.*]] = quake.alloca !quake.veq<1>
```
Comment (khalatepradnya) [inline test/Quake/cse.qke:87]: This test clearly shows two `concat` calls being optimized away...

File: test/Quake/cse.qke [hunk 3]
```diff
@@ -49,14 +58,84 @@ module {
...
+  return
+}
+
+// CHECK-LABEL:   func.func @canonicalize_multiple_concat() {
+// CHECK:           %[[VAL_0:.*]] = quake.alloca !quake.ref
+// CHECK:           %[[VAL_1:.*]] = quake.alloca !quake.ref
+// CHECK:           %[[VAL_2:.*]] = quake.concat %[[VAL_0]], %[[VAL_1]] : (!quake.ref, !quake.ref) -> !quake.veq<2>
+// CHECK:           %[[VAL_3:.*]] = quake.mz %[[VAL_2]] : (!quake.veq<2>) -> !cc.stdvec<!quake.measure>
+// CHECK:           %[[VAL_4:.*]] = quake.alloca !quake.veq<1>
+// CHECK:           %[[VAL_5:.*]] = quake.concat %[[VAL_0]], %[[VAL_1]], %[[VAL_4]] : (!quake.ref, !quake.ref, !quake.veq<1>) -> !quake.veq<3>
```
Comment (khalatepradnya) [inline test/Quake/cse.qke:127]: However, in this test, the `concat` calls are intact and order of `relax_size` is different than input. Can you add a clarifying comment as to what this test is checking?

File: test/Quake/cse.qke (hunk 1 above)
Comment (schweitzpgi) [reply test/Quake/cse.qke:16]: `module` is not needed and can be a small stumbling block to extending the test later.

File: test/Quake/cse.qke (hunk 3 above)
Comment (schweitzpgi) [reply test/Quake/cse.qke:127]: This test is a "negative test" to make sure the peephole isn't removing `concat` operations that cannot be removed.

It is also checking that `concat` is reified to a constant length whenever the sizes of its arguments are themselves constant. However, that reification may also imply that a `relax_size` (virtualizing the length) has to be reintroduced for type correctness.

**Review by khalatepradnya — APPROVED** [review]

**Pull request merged**